regcomp.c

```diff
@@ -138,6 +138,21 @@ bitset_on_num(BitSetRef bs)
 }
 #endif
 
+// Attempt to right size allocated buffers for a regex post compile
+static void
+onig_reg_resize(regex_t *reg)
+{
+    if (reg->alloc > reg->used) {
+      unsigned char *new_ptr = xrealloc(reg->p, reg->used);
+      // Skip the right size optimization if memory allocation fails
+      if (new_ptr) {
+        reg->alloc = reg->used;
+        reg->p = new_ptr;
+      }
+    }
+    if (reg->chain) onig_reg_resize(reg->chain);
```

Comment:
Do you expect all compilers to optimize tail-calls?

Reply:
Good point, I'd expect gcc and clang to with opt level 3, but we cannot rely on that. Or the behavior of other compilers. I replaced the recursive call with a goto instead in  https://github.com/ruby/ruby/pull/2696/commits/1ff3033fbba1be2aeda0cdafb5fe6b5a095059fa